test/state/host.hpp

```diff
@@ -0,0 +1,72 @@
+// evmone: Fast Ethereum Virtual Machine implementation
+// Copyright 2022 The evmone Authors.
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+
+#include "state.hpp"
+#include <optional>
+#include <unordered_set>
+
+namespace evmone::state
+{
+using evmc::uint256be;
+
+class Host : public evmc::Host
+{
+    evmc_revision m_rev;
+    evmc::VM& m_vm;
+    State& m_state;
+    const BlockInfo& m_block;
+    const Transaction& m_tx;
+    std::unordered_set<address> m_accessed_addresses;
+    std::vector<address> m_destructs;
+    std::vector<Log> m_logs;
+
+public:
+    Host(evmc_revision rev, evmc::VM& vm, State& state, const BlockInfo& block,
+        const Transaction& tx) noexcept
+      : m_rev{rev}, m_vm{vm}, m_state{state}, m_block{block}, m_tx{tx}
+    {}
+
+    [[nodiscard]] const auto& get_destructs() const noexcept { return m_destructs; }
+
+    [[nodiscard]] std::vector<Log>&& take_logs() noexcept { return std::move(m_logs); }
```

Comment:
Why is this moving while the others are just references?

Reply:
We return the logs in the end so we can move them out of the Host. The other structures are discarded.